iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/ArgDefinedOutsideOfScopeCheck.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * SonarQube IaC Plugin
+ * Copyright (C) 2021-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.checks;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.sonar.check.Rule;
+import org.sonar.iac.common.api.checks.CheckContext;
+import org.sonar.iac.common.api.checks.IacCheck;
+import org.sonar.iac.common.api.checks.InitContext;
+import org.sonar.iac.docker.tree.api.ArgInstruction;
+import org.sonar.iac.docker.tree.api.Body;
+import org.sonar.iac.docker.tree.api.DockerImage;
+import org.sonar.iac.docker.tree.api.Expression;
+import org.sonar.iac.docker.tree.api.Literal;
+import org.sonar.iac.docker.tree.api.RunInstruction;
+import org.sonar.iac.docker.tree.api.Variable;
+
+@Rule(key = "S6579")
+public class ArgDefinedOutsideOfScopeCheck implements IacCheck {
+  private static final String MESSAGE = "Include the ARG instruction in the build stage where it is used.";
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(Body.class, ArgDefinedOutsideOfScopeCheck::checkReferencedVariablesAreInScope);
+  }
+
+  private static void checkReferencedVariablesAreInScope(CheckContext ctx, Body body) {
+    List<String> globalArgsNames = collectArgNames(body.globalArgs().stream());
```

Comment:
I see what you are doing here.

It may not be necessary to do this, because in our docker analyzer, we have a Symbol table for variables.
Currently, thanks to the `DockerSymbolVisitor`, each variable has a Symbol associated, and this Symbol give you direct access to all reference of the variable.

Typically, the first reference is the "declaration" of the variable, and even better, it has a scope associated, which is either `Global` (`ARG` instruction at the beginning of the file, before any `FROM`) or `Image` (`ARG` instruction in a docker image, so after a `FROM` instruction). So exactly what you are looking for in your rule :smile: